ctest-next/templates/test.rs

```diff
@@ -0,0 +1,106 @@
+use std::mem;
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use std::fmt::{Debug, LowerHex};
+use std::ptr;
+use std::slice;
+use std::ffi::CStr;
+
+static FAILED: AtomicBool = AtomicBool::new(false);
+static NTESTS: AtomicUsize = AtomicUsize::new(0);
+
+fn main() {
+    eprintln!("RUNNING ALL TESTS");
+    run_all();
+    if FAILED.load(Ordering::Relaxed) {
+        panic!("some tests failed");
+    } else {
+        eprintln!("PASSED {} tests", NTESTS.load(Ordering::Relaxed));
+    }
+}
+
+/// Check that the value returned from the Rust and C side in a certain test is equivalent.
+///
+/// Internally it will remember which checks failed and how many tests have been run.
+fn check_same<T: PartialEq + Debug>(rust: T, c: T, attr: &str) {
+    if rust != c {
+        eprintln!("bad {}: rust: {:?} != c {:?}", attr, rust, c);
+        FAILED.store(true, Ordering::Relaxed);
+    } else {
+        NTESTS.fetch_add(1, Ordering::Relaxed);
+    }
+}
+
+/// Check that the value returned from the Rust and C side in a certain test is equivalent.
+///
+/// Internally it will remember which checks failed and how many tests have been run. This
+/// method is the same as `check_same` but prints errors in bytes in hex.
+fn check_same_hex<T: PartialEq + LowerHex + Debug>(rust: T, c: T, attr: &str) {
+    if rust != c {
+        let rust = format!("{:?} ({:#x})", rust, rust);
+        let c = format!("{:?} ({:#x})", c, c);
+        eprintln!("bad {}: rust: {} != c {}", attr, rust, c);
+        FAILED.store(true, Ordering::Relaxed);
+    } else {
+        NTESTS.fetch_add(1, Ordering::Relaxed);
+    }
+}
+
+
+{% for constant in ffi_items.constants() %}
+    {% let ty = constant.ty.to_token_stream().to_string() %}
+    {% let ident = constant.ident() %}
+
+    // Test that the string constant is the same in both Rust and C.
+    // While fat pointers can't be translated, we instead of * const c_char.
```

Comment:
Move these comments inside the `{% if %}`, since right now both comments look like they're getting instantiated each time.